File: 프로그래머스/3/43163. 단어 변환/단어 변환.js
function solution(begin, target, words) {
    if(!words.includes(target)) return 0;
    
    let map = new Map();
    
    for(let i=0; i<words.length-1; i++){
        for(let j=i+1; j<words.length; j++){
            const word1=words[i];
            const word2=words[j];
            let same=0;
            for(let i=0; i<word1.length; i++){
                if(word1[i]===word2[i]) same++;
            }
            if(word1.length-same===1){
                if (!map.has(word1)) map.set(word1, []);
                if (!map.has(word2)) map.set(word2, []);
                
                map.get(word1).push(word2);
                map.get(word2).push(word1);
            }
        }
    }
    
    let visited=new Set();
    let queue=[];
    for(const word of words){
        let same=0;
        for(let j=0; j<word.length; j++){
            if(word[j]===begin[j]) same++;
        }
        if(word.length-same===1) queue.push([word,1]);
    }
    
    while(queue.length){
        const [word, count]=queue.shift();
        if(word===target) return count;
        visited.add(word);
        for(const link of map.get(word)){
            if(!visited.has(link)) queue.push([link, count+1]);
        }
    }
    
}

File: 프로그래머스/3/43163. 단어 변환/단어 변환.java
import java.util.*;
class Solution {
    class Pair{
        String word;
        int n;

        Pair(String word, int n){
            this.word = word;
            this.n = n;
        }
    }
    
    public int solution(String begin, String target, String[] words) {
        List<String> list = new ArrayList<>(Arrays.asList(words));
        if(!list.contains(target)) return 0;
        list.add(begin);
        
        // 인접리스트
        Map<String, List<String>> map = new HashMap<>();
        
        for(String w1 : list){
            List<String> neighbor = new ArrayList<>();
            for(String w2 : list){
                if(w1.equals(w2)) continue;
                int cnt=0;
                for(int i=0; i<w1.length(); i++){
                    if(w1.charAt(i) != w2.charAt(i)) cnt++;
                }
                if(cnt==1) neighbor.add(w2);
            
            }
            map.put(w1, neighbor);
        }
        
        // BFS
        Queue<Pair> queue = new LinkedList<>();
        queue.offer(new Pair(begin, 0));
        
        Set<String> visited = new HashSet<>();
        visited.add(begin);
        
        while(!queue.isEmpty()){
            Pair pair = queue.poll();
            String curr = pair.word;
            
            for(String next : map.get(curr)){
                if(visited.contains(next)) continue;
                
                if(next.equals(target)) return pair.n+1;
                
                queue.offer(new Pair(next, pair.n+1));
                visited.add(next);
            }
        }
        return 0;
    }
}